File: bench/benches/bintrees.jl
include(joinpath(@__DIR__, "..", "lib", "bench.jl"))

# bintrees -- the benchmark-game binary-trees (see bench/benches/bintrees.l): a
# GC-throughput / long-lived-survival workload. build a stretch tree of depth
# max+1, hold a long-lived tree of depth max alive across the run, then for each
# depth d in min..max step 2 build 2^(max-d+min) short-lived trees and sum their
# node counts. a leaf is nothing and counts 0.
struct BNode
    l::Union{BNode,Nothing}    # TYPED fields: untyped (::Any) boxes every node and
    r::Union{BNode,Nothing}    # defeats inference -- the classic Julia perf trap.
end

mk(d) = d < 1 ? nothing : BNode(mk(d - 1), mk(d - 1))
ck(t) = t === nothing ? 0 : 1 + ck(t.l) + ck(t.r)

function bt_run(mn, mx)
    stretch = ck(mk(mx + 1))
    long = mk(mx)                       # LONG-LIVED -- survives the loop below
    total = 0
    for d in mn:2:mx
        n = 1 << (mx - d + mn)
        s = 0
        for _ in 1:n
            s += ck(mk(d))
        end
        total += s
    end
    return stretch + ck(long) + total
end

bench("bintrees", () -> bt_run(4, 14))

File: bench/benches/bintrees.rs
include!("../lib/bench.rs");

// bintrees -- the benchmark-game binary-trees (see bench/benches/bintrees.l): a
// GC-throughput / long-lived-survival workload. distinct heap-allocated nodes (Box):
// a malloc + an individual free PER NODE -- the honest object-churn every GC'd language
// does here, with rust's malloc/free instead of a collector. (we deliberately do NOT
// flatten to an index arena: that measures array indexing, not allocating/reclaiming
// distinct nodes -- the very thing this bench tests.) a leaf is None and counts 0.
enum Tree {
    Leaf,
    Node(Box<Tree>, Box<Tree>),
}

fn mk(d: i64) -> Tree {
    if d < 1 { Tree::Leaf } else { Tree::Node(Box::new(mk(d - 1)), Box::new(mk(d - 1))) }
}

fn ck(t: &Tree) -> i64 {
    match t {
        Tree::Leaf => 0,
        Tree::Node(l, r) => 1 + ck(l) + ck(r),
    }
}

fn bt_run(mn: i64, mx: i64) -> i64 {
    let stretch = ck(&mk(mx + 1));
    let long = mk(mx); // LONG-LIVED -- survives the loop below
    let mut total: i64 = 0;
    let mut d = mn;
    while d <= mx {
        let n: i64 = 1 << (mx - d + mn);
        let mut s: i64 = 0;
        for _ in 0..n {
            s += ck(&mk(d));
        }
        total += s;
        d += 2;
    }
    stretch + ck(&long) + total
}

fn main() {
    bench("bintrees", || bt_run(4, 14));
}
